net/instaweb/http/redirect_following_url_async_fetcher.cc

```diff
@@ -0,0 +1,294 @@
+    }
+    return true;
+  }
+
+  virtual void HandleHeadersComplete() {
+    GoogleString h = response_headers()->ToString();
+    // Currently we support permanent and temporary redirects.
+    recieved_redirect_status_code_ =
+        response_headers()->status_code() == HttpStatus::kMovedPermanently ||
+        response_headers()->status_code() == HttpStatus::kFound;
+
+    if (!recieved_redirect_status_code_) {
+      SharedAsyncFetch::HandleHeadersComplete();
+    }
+  }
+
+  virtual bool HandleWrite(const StringPiece& content,
+                           MessageHandler* handler) {
+    if (!recieved_redirect_status_code_) {
+      return SharedAsyncFetch::HandleWrite(content, handler);
+    }
+    return true;
+  }
+
+  virtual void HandleDone(bool success) {
+    DCHECK(gurl_.IsWebValid() || !success);
+    const RewriteOptions* options =
+        redirect_following_fetcher_->rewrite_options();
+
+    if (!recieved_redirect_status_code_) {
+      SharedAsyncFetch::HandleDone(success);
+      delete this;
+      return;
+    }
+
+    bool my_success = success;
+    int redirects_followed = urls_seen_->size();
+
+    if (my_success) {
+      // TODO(oschaaf): does this have the correct arguments?
+      bool cacheable = response_headers()->IsProxyCacheable(
+          base_fetch_->request_headers()->GetProperties(),
+          ResponseHeaders::GetVaryOption(options->respect_vary()),
+          ResponseHeaders::kHasValidator);
+      if (!cacheable) {
+        LOG(WARNING) << "Not following uncacheable redirect: " << url_;
+        my_success = false;
+      }
+    }
+
+    const char* raw_redirect_url =
+        response_headers()->Lookup1(HttpAttributes::kLocation);
+    GoogleString redirect_url;
+
+    if (!raw_redirect_url) {
+      LOG(WARNING) << "Failed looking up exactly one Location: header " << url_;
+      my_success = false;
+    } else if (!strlen(raw_redirect_url)) {
+      LOG(WARNING) << "Empty Location: header value";
+      my_success = false;
+    } else {
+      redirect_url = GoogleUrl::Sanitize(raw_redirect_url);
+      raw_redirect_url = redirect_url.c_str();
+    }
+
+    if (my_success && FindIgnoreCase(redirect_url, "#") != StringPiece::npos) {
+      LOG(WARNING) << "Decline redirect to " << redirect_url
+                   << " fragments are not supported.";
+      my_success = false;
+    }
+
+    scoped_ptr<GoogleUrl> redirect_gurl;
+    if (my_success) {
+      redirect_gurl.reset(new GoogleUrl(gurl_, redirect_url));
+      if (!redirect_gurl->IsWebValid()) {
+        LOG(WARNING) << "Invalid or unsupported url in location header: "
+                     << redirect_url;
+        my_success = false;
+      } else {
+        redirect_url = redirect_gurl->spec_c_str();
+      }
+    }
+
+    if (my_success) {
+      if (redirects_followed > redirect_following_fetcher_->max_redirects()) {
+        LOG(WARNING) << "Max redirects " << redirects_followed << " / "
```

Comment:
Do we want to have all these warnings echoed in a 404 page, rather than just putting them the server log?

Also it'd be better to carry a message_context_ so they go in the right server log.  E.g. in HTTPD there are vhost-specific loglevels.